polymetis/polymetis/tests/cpp/test_server.cpp

```diff
@@ -124,7 +124,7 @@ TEST_F(ServiceTest, SetController) {
                                       dummy_robot_state_, new TorqueCommand)
                       .ok());
 
-      if (i == 1) {
+      if (i == 2) {
```

Comment:
I'm slightly suspicious of the changes in `polymetis_server.cpp` because it causes this subtle off-by-1 here... My summary of the effect is that cleaning up a "TERMINATING" controller occurs in the _next_ call to ControlUpdate. Any thoughts on potentially dangerous downstream effects of the changes?

Reply:
The previous design cleans up a "TERMINATING" controller at the end of the current timestep, which includes the resetting of a the default controller. The problem with this is that if the controller status changes to TERMINATING before the controller selection step (either due to a untimely TerminateController call from the user, or due to stale controller state caused by a automatic restart), the controller would select and load a default controller that has not been reset to the current pose, which would have safety concerns.

Thus my solution is to process the TERMINATING controller _right_ before the controller selection step so that the default controller is always initialized when newly loaded. This is done by moving the step for cleaning up TERMINATING controllers up to the beginning of the loop, and also moving the controller selection step up to be right next to it. Could also parse the robot state before the two steps in question, it doesn't really matter.

The off-by-1 phenomena in the test is actually caused by the termination being one step __faster__. As you can see I increased the number of steps the controller was allowed to run by 1, while leaving the expected episode length untouched. This is because the ControlUpdate call after the TerminateController now finalizes the episode at the beginning of the loop, instead of loading & running the default controller and including it in the episode. I believe this is an improvement.

Reply:
> The problem with this is that if the controller status changes to TERMINATING before the controller selection step (either due to a untimely TerminateController call from the user, or due to stale controller state caused by a automatic restart), the controller would select and load a default controller that has not been reset to the current pose, which would have safety concerns.

This is definitely a safety issue, so thanks for catching it! From what I can tell, we don't actually guarantee thread-safety from untimely TerminateController calls simply by processing the TERMINATING controller right before the controller selection -- we'd need a mutex for that. In fact, we already have a mutex to make controller switching safe (`service_mtx_`).

I think the fundamental issue is that switching to a default controller is not thread safe in our implementation. How about we eliminate `TERMINATING` as a state in our controller finite-state-machine, and instead have a method which switches to & resets the default controller, which is protected by `service_mtx_`?

Reply:
I think using a mutex makes perfect sense here. 
However both internal and external terminations need to go through the same process of saving the episode indices in addition to resetting the default controller, which might cause race conditions with the controller loop.
Do you think it should be sufficient to simply protect the termination handling and controller selection part (the two blocks of code that I moved) with `service_mtx_`?